sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+### Table of contents
+
+- [SDK Refactor Design Proposal](#sdk-refactor-design-proposal)
+    + [Table of contents](#table-of-contents)
+    
+    + [Resources](#resources)
+    
+    + [Motivation](#motivation)
+    
+    + [Current SDK implementation](#current-sdk-implementation)
+      - [Flow](#flow)
+      
+    + [Limitations in current SDK Flow](#limitations-in-current-sdk-flow)
+        * [Client constructor](#client-constructor)
+        * [Proposed flow](#proposed-flow)
+        * [API and commands](#api-and-commands)
+        * [Proposed flow](#proposed-flow-1)
+        
+    + [Full proposed SDK flow](#full-proposed-sdk-flow)
+    
+    + [Backwards compatibility](#backwards-compatibility)
+    
+    + [Delivery plan](#delivery-plan)
+    
+        
+
+### Resources
+
+| Resource                  | Link                                      |
+| ------------------------- | ----------------------------------------- |
+| Zen of Python (the Bible) | https://www.python.org/dev/peps/pep-0020/ |
+
+
+
+### Motivation
+
+While GAing the CLI, it was realized that the SDK has the following limitations, making the code not easily readiable and hard to make additions:
+
+1. Classes handle multiple responsibilities, impeding on the ability to cover current and future usecases without expensive refactoring
+2. Implicit instead of explicit, forcing the end-user to understand the inner workings of our Client and the context they are running in, instead of allowing them to explicitly accomplish their desired flow. See [Zen of Python](https://www.python.org/dev/peps/pep-0020/) for more details.
+3. Areas of volatility are not encapsulated, becomes problematic when the requirements change or additions need to be made.
+
+In this document, I will go into more detail for each of these points as well as supply a proposal for how to address them. 
+
+It is important to note that this refactor will help us in our effort to decouple the CLI logic from the SDK. That way, when we are ready and decide to make that separation, we can do so more easily.
+
+
+
+### Current SDK implementation
+
+#### Flow
+
+The below image details the calls from when a user initalizes the *current* Client to when they receive a response.
+
+![CurrentSDKFlow](design/assets/CurrentSDKFlow.png)
+
+At a high level,
+
+1. The user initializes the *Client* with details needed to connect to the *Conjur server*
+
+    `client = conjur.Client(url="https://someserver", account="cucumber", password="123", ...)`.
+
+2. The *Client* constructor is called, ensuring all configuration and credential information needed to make a request have been passed in. If not, it fetches the missing information in a discovery-like manner. For example, for `login`, if a Keystore is not available, it will attempt to fetch credentials from the netrc. At this point the *API* is also initalized.
```

Comment:
What is the flow if partial arguments are passed by method? Is it still going into discovery mode?
Do we think about expanding it to environment variables like AWS CLI has?

Reply:
See the bottom under "Proposed flow" where I mention:

> We will *not* create the conjurrc and the netrc for the user as is true in the current state of the SDK. 

> We will *not* support partial states as is true in the current SDK. For example, we will not support the flow if a user provides `client.Configuration(action=file, url="https://someserver", account="", cert_file="/some/certfile")`. In this flow we will fail with relevant error.

I don't think we should go into discovery mode because it is an endless black hole. If a user provided username but not password, we would go into discovery mode and look for the password in the keystore and netrc. Let's say keystore has it and we find it there. Then realize they didn't provide url so again we perform this discovery again but this time it is _not_ in the keystore, rather in the netrc. So you see we have this weird state where we are pulling variables from different available locations

> Do we think about expanding it to environment variables like AWS CLI has?

Definitely on our radar! We won't be adding this functionality though as part of the refactor. In the refactor our goal is to think of the project and lay it out in a way that adding this functionality won't be hard